table.cc

```diff
@@ -89,8 +89,9 @@ table::make_sstable_reader(schema_ptr s,
                     tracing::trace_state_ptr trace_state,
                     streamed_mutation::forwarding fwd,
                     mutation_reader::forwarding fwd_mr) {
+                assert(pr.is_singular() && pr.start()->value().has_key());
```

Comment:
This `assert()` is completely redundant, we have an if above checking this very condition.

Reply:
It's not immediately obvious from the code that this `pr` is the same `pr` as above (this `pr` comes from the lambda wrapped by `mutation_source`, the previous `pr` which was checked comes from an outer lambda). In fact it's not obvious for me at all, I assumed it must be true, but it's better to check.

Reply:
I don't think this is justified here, as we are in the appropriate branch of an if statement conditioned on the very same condition.

Reply:
I'm trying to argue that it's not the same condition. The previous check was performed on a different variable. It's easier to see if we rename the two:
```
table::make_sstable_reader(...
                                   const dht::partition_range& pr1,
                                   ...) const {
    auto ms = [&] () -> mutation_source {
        if (pr1.is_singular() && pr1.start()->value().has_key()) {
            ...
            return mutation_source([...] (
                    ...
                    const dht::partition_range& pr2,
                    ...) {
                assert(pr2.is_singular() && pr2.start()->value().has_key());
```

it is not clear from the code that `pr1` is the same as `pr2`. The condition was on `pr1` and the assert is on `pr2`.

I don't get why it's done that way, i.e. why the parameters are passed through two different paths.

Reply:
It is the same. The `mutation_source` will be passed `pr1`.

Reply:
Ok, then I will remove the assert.

But couldn't the code be somehow refactored so that it's enforced in compile-time and obvious by static analysis? Why does the lambda inside `mutation_source` have to take all those parameters again?

Reply:
We decide on the `mutation_source` to return based on whether the range is singular or not. We could just as well return a single `mutation_source` which decideds on the reader to create based on the range passed to it. I might be less confusing.

Reply:
So we could do
```
table::make_sstable_reader(... params ...) {
    auto ms = [] () { return  mutation_source([] (... params ...) {
        ... branch here ...
    }};
    return make_restricted_flat_reader(std::move(ms), ...);
}
```

Reply:
Yes, except the outer lambda is now unnecessary.